src/libraries/System.Drawing.Common/src/System/Drawing/Graphics.cs

```diff
@@ -64,26 +65,32 @@ public sealed class Graphics : MarshalByRefObject, IDisposable, IDeviceContext
         public delegate bool DrawImageAbort(IntPtr callbackdata);
 
 #if NET7_0_OR_GREATER
-        [CustomTypeMarshaller(typeof(DrawImageAbort), CustomTypeMarshallerKind.Value, Direction = CustomTypeMarshallerDirection.In, Features = CustomTypeMarshallerFeatures.TwoStageMarshalling | CustomTypeMarshallerFeatures.UnmanagedResources)]
-        internal unsafe struct DrawImageAbortMarshaller
+        [CustomMarshaller(typeof(DrawImageAbort), MarshalMode.ManagedToUnmanagedIn, typeof(KeepAliveMarshaller))]
+        internal static class DrawImageAbortMarshaller
         {
-            private delegate Interop.BOOL DrawImageAbortNative(IntPtr callbackdata);
-            private DrawImageAbortNative? _managed;
-            private delegate* unmanaged<IntPtr, Interop.BOOL> _nativeFunction;
-            public DrawImageAbortMarshaller(DrawImageAbort? managed)
+            internal unsafe struct KeepAliveMarshaller
             {
-                _managed = managed is null ? null : data => managed(data) ? Interop.BOOL.TRUE : Interop.BOOL.FALSE;
-                _nativeFunction = _managed is null ? null : (delegate* unmanaged<IntPtr, Interop.BOOL>)Marshal.GetFunctionPointerForDelegate(_managed);
-            }
+                private delegate Interop.BOOL DrawImageAbortNative(IntPtr callbackdata);
+                private DrawImageAbortNative? _managed;
+                private delegate* unmanaged<IntPtr, Interop.BOOL> _nativeFunction;
+                public void FromManaged(DrawImageAbort? managed)
+                {
+                    _managed = managed is null ? null : data => managed(data) ? Interop.BOOL.TRUE : Interop.BOOL.FALSE;
+                    _nativeFunction = _managed is null ? null : (delegate* unmanaged<IntPtr, Interop.BOOL>)Marshal.GetFunctionPointerForDelegate(_managed);
+                }
 
-            public delegate* unmanaged<IntPtr, Interop.BOOL> ToNativeValue()
-            {
-                return _nativeFunction;
-            }
+                public delegate* unmanaged<IntPtr, Interop.BOOL> ToUnmanaged()
+                {
+                    return _nativeFunction;
+                }
 
-            public void FreeNative()
-            {
-                GC.KeepAlive(_managed);
+                public void OnInvoked()
+                {
+                    GC.KeepAlive(_managed);
+                }
+
+                [SuppressMessage("Performance", "CA1822:Mark members as static", Justification = "This method is part of the marshaller shape and is required to be an instance method.")]
+                public void Free() { }
```

Comment:
Here's how the empty `Free` methods look when they're required on the stateful shape. Just wanted to raise it for us to look at.

Might be worth investigating introducing a DiagnosticSuppressor to suppress this diagnostic in these cases as following the code fix would break the marshaller (and the diagnostic is configured to be an error in dotnet/runtime, so it needs to be manually suppressed here today)

Reply:
Oof, yeah, that is unfortunate. We should look into a suppressor - can you file an issue?

Reply:
Will do.